src/libmongoc/src/mongoc/mongoc-handshake.c

```diff
@@ -628,29 +623,37 @@ mongoc_handshake_data_append (const char *driver_name,
       return false;
    }
 
+   BSON_ASSERT (_mongoc_handshake_get ()->platform);
+
    /* allow practically any size for "platform", we'll trim it down in
```

Comment:
Ok, it looks like this is idiomatic WRT how handshakes are assembled, and explains the frequent bson_free() calls, rather than just overwriting into already-allocated memory.